fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,783 @@
+/// This codec takes a large number space (u128) and reduces it to a compact number space.
+///
+/// It will find spaces in the numer range. For example:
+///
+/// 100, 101, 102, 103, 104, 50000, 50001
+/// could be mapped to
+/// 100..104 -> 0..4
+/// 50000..50001 -> 5..6
+///
+/// Compact space 0..=6 requires much less bits than 100..=50001
+///
+/// The codec is created to compress ip addresses, but may be employed in other use cases.
+use std::{
+    cmp::Ordering,
+    collections::BinaryHeap,
+    io::{self, Write},
+    net::{IpAddr, Ipv6Addr},
+    ops::RangeInclusive,
+};
+
+use common::{BinarySerializable, CountingWriter, VIntU128};
+use ownedbytes::OwnedBytes;
+use tantivy_bitpacker::{self, BitPacker, BitUnpacker};
+
+use crate::column::{ColumnV2, ColumnV2Ext};
+
+pub fn ip_to_u128(ip_addr: IpAddr) -> u128 {
+    let ip_addr_v6: Ipv6Addr = match ip_addr {
+        IpAddr::V4(v4) => v4.to_ipv6_mapped(),
+        IpAddr::V6(v6) => v6,
+    };
+    u128::from_be_bytes(ip_addr_v6.octets())
+}
+
+const INTERVAL_COST_IN_BITS: usize = 64;
+
+/// Store blank size and position. Order by blank size.
+///
+/// A blank is an unoccupied space in the data.
+/// E.g. [100, 201] would have a `BlankAndPos{ pos: 0, blank_size: 100}`
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+struct BlankSizeAndPos {
+    blank_size: u128,
+    /// Position in the sorted data.
+    pos: usize,
+}
+impl BlankSizeAndPos {
+    fn new(blank: u128, pos: usize) -> Self {
+        BlankSizeAndPos {
+            blank_size: blank,
+            pos,
+        }
+    }
+}
+
+impl Ord for BlankSizeAndPos {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.blank_size.cmp(&other.blank_size)
+    }
+}
+impl PartialOrd for BlankSizeAndPos {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.blank_size.partial_cmp(&other.blank_size)
+    }
+}
+
+#[test]
+fn test_delta_and_pos_sort() {
+    let mut deltas: BinaryHeap<BlankSizeAndPos> = BinaryHeap::new();
+    deltas.push(BlankSizeAndPos::new(10, 1));
```

Comment:
if the goal is to test Ord, let's test Ord. What is the point of the Binary heap here?

Reply:
I wanted to test that `heap.pop` gives descending blank_size. renamed the test